### PR TITLE
[PT-167808593] Improve state channel systems tests

### PR DESCRIPTION
**Pull request by grzegorz225**
System tests now test:
- Computability of off-chain transfers, contract deployment and contract calls inside a state channel between different Fortuna releases.
- Compatibility of leave/reestablish between different Fortuna releases.
- Compatibility of the off chain state channel database after upgrading the node version(The tests for this are commented out because currently all Fortuna releases lose the off-chain state after a while - the first release without this problem would be 4.2.0 as the fix for this was merged today).

PT: https://www.pivotaltracker.com/story/show/167808593

### CODE REVIEW

**Comment by velzevur** [comment]
I've pushed this branch to `system-tests`. This will run the [system tests](https://circleci.com/gh/aeternity/aeternity/59290?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) in CI

File: system_test/common/aest_db_SUITE.erl [hunk 1]
```diff
@@ -18,7 +18,11 @@
          node_can_reuse_db_of_minerva_node_with_epoch_db/1,
          minerva_node_with_channels_update_as_tuple_can_reuse_db_of_analogous_node/1,
          minerva_node_with_channels_update_as_tuple_can_reuse_db_of_analogous_node_with_force_progress_tx/1,
-         node_can_reuse_db_of_minerva_node_with_channels_update_as_tuple_with_force_progress_tx/1
+         node_can_reuse_db_of_minerva_node_with_channels_update_as_tuple_with_force_progress_tx/1,
+         %% TODO: Shouldn't the names be longer ;)
```
Comment (velzevur) [inline system_test/common/aest_db_SUITE.erl]: this is a long discussion....

File: system_test/common/aest_db_SUITE_data/aeternity.yaml.mustache [hunk 2]
```diff
@@ -1,6 +1,24 @@
...
+sync:
+    port: {{services.sync.port}}
+    single_outbound_per_group: false
+
+websocket:
+    channel:
+        listen_address: 0.0.0.0
+        port: {{services.ext_ws.port}}
+
+keys:
```
Comment (velzevur) [inline system_test/common/aest_db_SUITE_data/aeternity.yaml.mustache]: Do we need this?

**Comment by lucafavatella** [comment]
> Probably it is worth rebasing on top of master (with a [merged fix](https://github.com/aeternity/aeternity/pull/2641) for accidental SC state deletion)

I see this PR is already merged on latest master.

**Comment by lucafavatella** [comment]
I see this branch was pushed to the branch for running all system tests - good. I notice there is a failure though:
https://59290-99802036-gh.circle-artifacts.com/0/home/circleci/aeternity/system_test/logs/ct_run.aeternity_ct@localhost.2019-08-13_21.10.55/system_test.common.logs/run.2019-08-13_21.10.56/aest_channels_suite.init_per_suite.html
```
=== Reason: no match of right hand side value {error,
                                                    {compiler_error,
                                                     "/home/circleci/aeternity/_build/system_test+test/lib/aecontract/../../extras/test/contracts/identity.aes",
                                                     <<"Compiler error">>}}
  in function  aest_channels_SUITE:precompile_identity_contract/1 (/home/circleci/aeternity/_build/system_test+test/extras/system_test/common/aest_channels_SUITE.erl, line 239)
  in call from test_server:ts_tc/3 (test_server.erl, line 1539)
  in call from test_server:run_test_case_eval1/6 (test_server.erl, line 1140)
  in call from test_server:run_test_case_eval/9 (test_server.erl, line 987)
```

File: system_test/common/aest_channels_SUITE.erl [hunk 3]
```diff
@@ -221,8 +225,64 @@ simple_channel_test(ChannelOpts, InitiatorNodeBaseSpec, ResponderNodeBaseSpec, C
...
+
+%=== HELPERS ===================================================================
+
+precompile_identity_contract(Config1) ->
+    %% Precompile a simple contract for testing
+    Config2 = aect_test_utils:init_per_group(aevm, Config1),
+    aect_test_utils:setup_testcase(Config2),
+    SimpleContractName = "identity",
+    {ok, BinSrc} = aect_test_utils:read_contract(aect_test_utils:sophia_version(), SimpleContractName),
+    {ok, Code}   = aect_test_utils:compile_contract(aect_test_utils:sophia_version(), SimpleContractName),
```
Comment (lucafavatella) [inline system_test/common/aest_channels_SUITE.erl]: This line fails.

**Comment by grzegorz225** [comment]
> Probably it is worth rebasing on top of master (with a [merged fix](https://github.com/aeternity/aeternity/pull/2641) for accidental SC state deletion) and uncomenting tests

I will uncomment the tests after the 4.2.0 release is out - The tests require a docker container with the fix already present, otherwise I would need to make some large changes in order to compare the behavior between the two different commits.

File: system_test/common/aest_channels_SUITE.erl (hunk 3 above)
Comment (grzegorz225) [reply system_test/common/aest_channels_SUITE.erl]: I'm looking into it.

**Comment by grzegorz225** [comment]
System tests finally pass inside CI. This PR is blocked until a docker image with the 4.2.0 version is released - this is needed to uncomment the failing leave-upgrade-reestablish tests.

**Comment by grzegorz225** [comment]
> Is the problem not that you should do a "./rebar3 as system_test, test upgrade" in addition to the " as system_test" upgrade?

It might be, but currently I won't be able to check it in CI as the cache got updated with my workaround.

**Pull request merged**